File: Data Structures & Algorithms/maximum-subarray/submission-1.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int maxSub = nums[0];
        int currMax = nums[0];
        if (nums.size() == 1) {
            return nums[0];
        }
        for (int i = 1; i < nums.size(); i++) {
            if (currMax < 0) {
                currMax = 0;
            }
            currMax += nums[i];
            if (currMax > maxSub) {
                maxSub = currMax;
            }
        }
        return maxSub;
    }
};

File: Data Structures & Algorithms/maximum-product-subarray/submission-1.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = max(nums)
        currMax, currMin = 1, 1
        for num in nums:
            if num == 0:
                currMax, currMin = 1, 1
                continue
            tmp = currMax * num
            currMax = max(currMax * num, currMin * num, num)
            currMin = min(tmp, currMin * num, num)
            res = max(res, currMax, currMin)
        return res